src/ConstructiveSolidGeometry/SurfacePrimitives/ConeMantle.jl

```diff
@@ -1,150 +1,330 @@
+    ints2 = obj_l.origin + λ2 * obj_l.direction 
+    return _transform_into_global_coordinate_system(ints1, cm), 
+           _transform_into_global_coordinate_system(ints2, cm)
 end
 
-function ConeMantle(;rbot = 1, rtop = 0, φMin = 0, φMax = 2π, zMin = -1/2, zMax = 1/2)
-    T = float(promote_type(typeof.((rbot, rtop, φMin, φMax, zMin, zMax))...))
-    r = rbot == rtop ? T(rbot) : (T(rbot), T(rtop))
-    φ = mod(T(φMax) - T(φMin), T(2π)) == 0 ? nothing : T(φMin)..T(φMax)
-    z = zMax == -zMin ? T(zMax) : T(zMin)..T(zMax)
-    ConeMantle(T, r, φ, z)
-end
-ConeMantle(rbot, rtop, φMin, φMax, zMin, zMax) = ConeMantle(;rbot = rbot, rtop = rtop, φMin = φMin, φMax = φMax, zMin = zMin, zMax = zMax)
 
-function ConeMantle(rbot::R1, rtop::R2, height::H) where {R1<:Real, R2<:Real, H<:Real}
-    T = float(promote_type(R1, R2, H))
-    ConeMantle( T, (T(rbot), T(rtop)), nothing, T(height)/2)
+# x = term1/(2 term2) 
+
+# and D1 = -sqrt(D3^2 S^2 - D2^2) 
+# and -L1 sqrt(D3^2 S^2 - D2^2) + D2 L2 - D3 hZ S^2 - D3 L3 S^2 - D3 R0 S!=0
+
+"""
+    intersection(cm::ConeMantle{T,T}, l::Line{T}) where {T}
+
+The function will always return 2 CartesianPoint's.
+If the line just touches the mantle, the two points will be the same. 
+If the line does not touch the mantle at all, the two points will have NaN's as there coordinates.
+"""
+function intersection(cm::ConeMantle{T,T}, l::Line{T}) where {T}
+    obj_l = _transform_into_object_coordinate_system(l, cm) # direction is not normalized
+    
+    L1 = obj_l.origin.x
+    L2 = obj_l.origin.y
+    L3 = obj_l.origin.z
+    D1 = obj_l.direction.x
+    D2 = obj_l.direction.y
+    D3 = obj_l.direction.z
+
+    f1 = D1^2 + D2^2 
+    λ = inv(f1) # f1 is only 0 if obj_l is parallel to the axis of the cone 
+                # (here eZ -> D1 = D2 = 0)
+                # We assume here that this is not the case -> 
+                # We check this in choosing the sample / evaluating dimensions in `paint!` 
+    hZ = cm.hZ
+    R0 = cm.r
+
+    term1 = (2D1*L1 + 2D2*L2)^2
+    term2 = L1^2 + L2^2 - R0^2
+    term3 = -D1*L1 - D2*L2 
+    term4 = term1 - 4*f1*term2
+    sq::T = term4 < 0 ? T(NaN) : sqrt(term1 - 4*f1*term2) # if this 
+    
+    λ1 = λ * (-sq/2 + term3) 
+    λ2 = λ * (+sq/2 + term3)
+    
+    ints1 = obj_l.origin + λ1 * obj_l.direction 
+    ints2 = obj_l.origin + λ2 * obj_l.direction 
+    return _transform_into_global_coordinate_system(ints1, cm), 
+           _transform_into_global_coordinate_system(ints2, cm)
 end
 
 
-get_r_at_z(c::ConeMantle{T, T}, z::Real) where {T} = c.r
-get_r_at_z(c::ConeMantle{T, Tuple{T,T}}, z::Real) where {T} = _get_r_at_z(c.r[1], c.r[2], c.z, z)
+get_top_ellipse(cm::ConeMantle{T,T,Nothing}) where {T} = 
+    Ellipse(cm.r, cm.φ, cm.origin + cm.rotation * CartesianPoint(zero(T), zero(T), cm.hZ),  cm.rotation)
+get_bot_ellipse(cm::ConeMantle{T,T,Nothing}) where {T} = 
+    Ellipse(cm.r, cm.φ, cm.origin - cm.rotation * CartesianPoint(zero(T), zero(T), cm.hZ), RotZ(π) * -cm.rotation)
```

Comment:
Why `RotZ(π) * -cm.rotation` and not just `cm.rotation`? To rotate the `normal` vector?

Reply:
I am not sure whats the best way here. 
I used `RotZ(π) * -cm.rotation` that the edges between the ellipses in the plot recipe for a tube are nicely align. 
But I believe this is not important right now. But may get important when we want to use proper this in the drift.